### PR TITLE
WorkbenchErrorHandler: Avoid Display.getDefault()

### LINKED ISSUE #409: Deadlock when WorkbenchErrorHandler is called after platform shutdown was initiated
When the platform is shutting down,  the Display is disposed and thrown away.

However, exceptions may still be thrown in worker threads and `WorkbenchErrorHandler` is called to handle them.

`WorkbenchErrorHandler` then uses `Display.getDefault()` to get the Display. However, this violates the recommendation of `IWorkbench#getDisplay()` that
```
* Code should always ask the workbench for the display rather than rely on {@link Display#getDefault Display.getDefault()}.
```

Calling `Display.getDefault()` will implicitly attempt to create a *new* instance, which potentially deadlocks:

```
"main" id=1 state=WAITING waitedCount=70 blockedCount=135
	-- waiting for lock=java.util.concurrent.atomic.AtomicReference@e5d008b
	at java.base@11.0.16.1/java.lang.Object.wait(Native Method)
	at org.eclipse.osgi.container.SystemModule.waitForStop(SystemModule.java:173)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle.waitForStop(EquinoxBundle.java:312)
	at org.eclipse.osgi.launch.Equinox.waitForStop(Equinox.java:217)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.shutdown(EclipseStarter.java:458)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:274)
	at java.base@11.0.16.1/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.16.1/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.16.1/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.16.1/java.lang.reflect.Method.invoke(Method.java:566)
	at app//org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at app//org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at app//org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at app//org.eclipse.equinox.launcher.Main.main(Main.java:1440)

"Worker-5: Animation start" id=50 state=RUNNABLE waitedCount=109 blockedCount=86 eclipseJob=Animation start(29)
	at org.eclipse.swt.internal.gtk.GDK.gdk_threads_enter(Native Method)
	at org.eclipse.swt.widgets.Display.createDisplay(Display.java:1145)
	at org.eclipse.swt.widgets.Display.create(Display.java:1078)
	at org.eclipse.swt.graphics.Device.<init>(Device.java:168)
	-- locks java.lang.Class@4f4c789f
	at org.eclipse.swt.widgets.Display.<init>(Display.java:630)
	at org.eclipse.swt.widgets.Display.<init>(Display.java:621)
	at org.eclipse.swt.widgets.Display.getDefault(Display.java:2347)
	-- locks java.lang.Class@4f4c789f
	at org.eclipse.ui.statushandlers.WorkbenchErrorHandler.handle(WorkbenchErrorHandler.java:54)
	at org.eclipse.ui.internal.ide.IDEWorkbenchErrorHandler.handle(IDEWorkbenchErrorHandler.java:106)
	at org.eclipse.ui.internal.WorkbenchErrorHandlerProxy.handle(WorkbenchErrorHandlerProxy.java:34)
	at org.eclipse.ui.statushandlers.StatusManager.handle(StatusManager.java:217)
	at org.eclipse.ui.internal.progress.ProgressManager$1.done(ProgressManager.java:467)
	at org.eclipse.core.internal.jobs.JobListeners$$Lambda$159/0x00000008002e8040.notify(Unknown Source)
	at org.eclipse.core.internal.jobs.JobListeners.doNotify(JobListeners.java:73)
	at org.eclipse.core.internal.jobs.JobListeners.done(JobListeners.java:118)
	at org.eclipse.core.internal.jobs.JobManager.endJob(JobManager.java:718)
	at org.eclipse.core.internal.jobs.WorkerPool.endJob(WorkerPool.java:117)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:83)

"Framework stop - Equinox Container: 4a6271b8-3e80-40ed-bfa2-0b426652758c" id=69 state=BLOCKED waitedCount=0 blockedCount=1
	-- waiting for lock=java.lang.Class@4f4c789f lock_owner=50
	at org.eclipse.swt.widgets.Display.findDisplay(Display.java:1458)
	at org.eclipse.swt.widgets.Display.getCurrent(Display.java:1798)
	at com.sap.adt.tools.core.ui.internal.AdtToolsUiStartup.dispose(AdtToolsUiStartup.java:168)
	at com.sap.adt.tools.core.ui.Activator.stop(Activator.java:583)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:875)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.base@11.0.16.1/java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.stop(BundleContextImpl.java:867)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.stopWorker0(EquinoxBundle.java:1046)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.stopWorker(EquinoxBundle.java:376)
	at org.eclipse.osgi.container.Module.doStop(Module.java:660)
	at org.eclipse.osgi.container.Module.stop(Module.java:521)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.decStartLevel(ModuleContainer.java:1893)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1768)
	-- locks java.lang.Object@64e92f08
	at org.eclipse.osgi.container.SystemModule.stopWorker(SystemModule.java:275)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule.stopWorker(EquinoxBundle.java:208)
	at org.eclipse.osgi.container.Module.doStop(Module.java:660)
	at org.eclipse.osgi.container.Module.stop(Module.java:521)
	at org.eclipse.osgi.container.SystemModule.stop(SystemModule.java:207)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule$1.run(EquinoxBundle.java:226)
	at java.base@11.0.16.1/java.lang.Thread.run(Thread.java:829)
	Java locks:
		java.util.concurrent.locks.ReentrantLock$NonfairSync@7456905c
		java.util.concurrent.locks.ReentrantLock$NonfairSync@43926e92

```

Solution:
`WorkbenchErrorHandler` should use `PlatformUI.getWorkbench().getDisplay();` rather than `Display.getDefault()` to avoid creation of a new instance.

### PR DESCRIPTION
**Pull request by sratz**
WorkbenchErrorHandler might be called from worker threads after platform shutdown has already been initiated and the Display has already been disposed.

Do not use Display.getDefault(), since that may implicitly attempt to create a new instance of Display, which can lead to a deadlock in the worst case.

Instead, use PlatformUI.getWorkbench().getDisplay() to get a potential Display just like IWorkbench#getDisplay() recommends.

Fixes #409

### CODE REVIEW

File: bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/statushandlers/WorkbenchErrorHandler.java [hunk 1]
```diff
@@ -51,7 +53,17 @@ public void handle(final StatusAdapter statusAdapter, int style) {
 			if (Display.getCurrent() != null) {
 				showStatusAdapter(statusAdapter, block);
 			} else {
-				Display.getDefault().asyncExec(() -> showStatusAdapter(statusAdapter, block));
+				try {
+					// Checking isDisposed() here isn't enough due to race conditions
+					// since we are in a worker thread.
```
Comment (laeubi) [inline bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/statushandlers/WorkbenchErrorHandler.java:69]: You could check first with PlatformUI.isPlatformRunning()

Comment (sratz) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/statushandlers/WorkbenchErrorHandler.java:69]: That wouldn't be enough. There's a race-condition: Right after checking the platform might be shut down / the Display disposed, so the check won't really help.

Comment (sratz) [reply bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/statushandlers/WorkbenchErrorHandler.java:69]: I force-pushed and improved things a bit in the latest version.

**Review by laeubi — APPROVED** [review]
Looks good to me.

**Pull request merged**